### PR TITLE
Big refactor - extract all common code to warehouse

### PR DESCRIPTION
**Pull request by solidovic**
## Description

Extract all common code like `App layout`, `Header`, `Footer`, `Wallet modal` etc to warehouse (related PR is https://github.com/lidofinance/warehouse/pull/111).

Added a pseudo functional components like `FutureTxInfo` (for demonstration a some UI processes).

Added some API methods for demonstrations best practices (for example: `rateLimit`, `responseTimeMetric`, `cacheControl`, `defaultErrorHandler`) for Lido UI projects.

~~Pay no attention to `.yalc` and `file:.yalc/@lidofinance/*` in `package.json`, 'cause it will be removed after merged (https://github.com/lidofinance/warehouse/pull/111).~~

## Local testing
~~Needs to install [yalc](https://github.com/wclr/yalc) globally.
All new not published packages you can find in `.yalc` (more detail you can find here https://github.com/lidofinance/warehouse/pull/111)~~
**Not actual!**

## Motivation
- Easy updates of all widgets and related projects.
- Following principle `Don’t repeat yourself`.

fixes UI-601; fixes UI-545

### CODE REVIEW

File: .yalc/@lidofinance/next-api-wrapper/README.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
+# @lidofinance/next-api-wrapper
```
Comment (alx-khramov) [inline .yalc/@lidofinance/next-api-wrapper/README.md:1]: You forgot to add .yalc to .gitignore

**Review by alx-khramov — CHANGES_REQUESTED** [review]

File: yalc.lock [hunk 2]
```diff
@@ -0,0 +1,26 @@
+{
```
Comment (alx-khramov) [inline yalc.lock:1]: And this should be in .gitignore too

File: .env [hunk 3]
```diff
@@ -10,12 +10,26 @@ API_PROVIDER_URL_1=
 API_PROVIDER_URL_4=
 API_PROVIDER_URL_5=
 
+# Subgraph endpoint
+SUBGRAPH_MAINNET=https://api.thegraph.com/subgraphs/name/lidofinance/lido
+SUBGRAPH_ROPSTEN=
```
Comment (DiRaiks) [inline .env:15]: I think we need to shorten the list of supported chains. Not all of them are already working.

File: components/stakeForm/stakeForm.tsx [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  const openTxModal = useCallback(() => {
+    setIsTxModalOpen(true);
+  }, []);
+
+  const closeTxModal = useCallback(() => {
+    setIsTxModalOpen(false);
+  }, []);
+
+  const [txStage, setTxStage] = useState(TX_STAGE.IDLE);
+  // /Modals
```
Comment (DiRaiks) [inline components/stakeForm/stakeForm.tsx:38]: Empty content after comment

File: consts/estimate.ts [hunk 5]
```diff
@@ -0,0 +1,4 @@
+// Account for gas estimation
+// will always have >=0.001 ether, >=0.001 stETH, >=0.001 wstETH
+// on Mainnet, Rinkeby, Goerli
+export const ESTIMATE_ACCOUNT = '0x87c0e047F4e4D3e289A56a36570D4CB957A37Ef1';
```
Comment (DiRaiks) [inline consts/estimate.ts:4]: I think you shouldn't use such things in the template

File: components/stakeForm/stakeForm.tsx (hunk 4 above)
Comment (solidovic) [reply components/stakeForm/stakeForm.tsx:38]: // /Something - closes the comment block for the code

// Start comment block
```code```
// /closes the comment block

File: .env (hunk 3 above)
Comment (solidovic) [reply .env:15]: More context pls

File: .yalc/@lidofinance/next-api-wrapper/README.md (hunk 1 above)
Comment (solidovic) [reply .yalc/@lidofinance/next-api-wrapper/README.md:1]: Not forgotten. I will delete `.yalc` after packages will be published

File: yalc.lock (hunk 2 above)
Comment (solidovic) [reply yalc.lock:1]: above

File: .env (hunk 3 above)
Comment (solidovic) [reply .env:15]: only mainnet and goerli

File: consts/estimate.ts (hunk 5 above)
Comment (solidovic) [reply consts/estimate.ts:4]: think

File: hooks/useGasPrice.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+      return;
+    }
+
+    const staticProvider = getStaticRpcBatchProvider(
+      chainId as CHAINS,
+      getBackendRPCPath(chainId as CHAINS),
+    );
+
+    try {
+      const newGasPrice = await staticProvider.getGasPrice();
```
[the remark this answers is not in the document]
Comment (solidovic) [reply hooks/useGasPrice.ts:25]: think

File: utilsApi/getStEthApr.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  getOracleContractFactory,
+  getStethAddress,
+  getStethContractFactory,
+} from 'consts';
+
+import { serverLogger } from './serverLogger';
+import { rpcUrls } from './rpcUrls';
+import { getStaticRpcBatchProvider } from './rpcProviders';
+
+export const getStEthApr = async (): Promise<string> => {
```
[the remark this answers is not in the document]
Comment (solidovic) [reply utilsApi/getStEthApr.ts:16]: think or use service eth-api

File: HOW_TO.md [hunk 8]
```diff
@@ -0,0 +1,167 @@
+# HOW TO UPGRADE OR MAKE MIGRATION
+
+Current Lido CI setup is built with this app template in mind.
+If your project was built with custom code / fork of an older version of the staking widget,
+most likely there would be changes required to make your app run smoothly.
+
+### Step 0. Docs
+
+READ FIRST [DOCS.md](./DOCS.md)!!!
```
Comment (Jabher) [inline HOW_TO.md:9]: it sounds very aggressive now
```suggestion
First, read [DOCS.md](./DOCS.md)
```

File: HOW_TO.md [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+- `Header`,
+- `Main` - page content block,
+- `Footer`,
+- etc.
+
+More info [here](https://github.com/lidofinance/warehouse/tree/main/packages/next/widget-layout).
+
+If you need a completely different layout structure - see block `Not EVM widget` below.
+
+#### Not EVM widget
```
Comment (Jabher) [inline HOW_TO.md:74]:
```suggestion
#### Non-EVM widget
```

File: components/ethWalletCard/types.ts [hunk 10]
```diff
@@ -0,0 +1,3 @@
+import { WalletCardComponent } from '@lidofinance/ui-primitives';
+
+export type EthWalletCardComponent = WalletCardComponent;
```
Comment (Jabher) [inline components/ethWalletCard/types.ts:3]: is there a reason for that?

File: components/headerActions/headerActions.tsx [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+} from '@lidofinance/eth-ui-wallet-modal';
+import { useWeb3 } from 'reef-knot/web3-react';
+
+import { HeaderWalletChainStyle } from './styles';
+
+export const HeaderActions: FC = () => {
+  const { active, chainId } = useWeb3();
+
+  const chainName = chainId ? CHAINS[chainId] : CHAINS.Mainnet;
+  const testNet = chainId !== CHAINS.Mainnet;
```
Comment (Jabher) [inline components/headerActions/headerActions.tsx:15]: it is not always testnet. e.g. Polygon will match with it. please rename

File: config/matomoWalletsEvents.ts [hunk 12]
```diff
@@ -234,43 +226,38 @@ export const walletsMetrics: WalletsMetrics = {
...
+        // This code returns:
+        // onClickAmbire: bindEvent('onClickAmbire'),
+        // onClickBlockchaincom: bindEvent('onClickBlockchaincom'),
+        // more...
+        ...(
+          Object.keys(MATOMO_WALLETS_EVENTS_TYPES) as Array<
+            keyof typeof MATOMO_WALLETS_EVENTS_TYPES
+          >
+        ).reduce<Record<string, () => void>>((accumulator, eventKey) => {
+          if (eventKey.indexOf('onClick') > -1) {
```
Comment (Jabher) [inline config/matomoWalletsEvents.ts:238]: Alex fixed this behaviour, there is new event for all clicks, please consult with him

File: hooks/useTxCostInUsd.ts [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+  // useEthPrice hook works via mainnet chain!
+  const { data: ethInUsd } = useEthPrice() as {
+    data?: number;
+    update: () => void;
+  };
+
+  const [txCostInUsd, setTxCostInUsd] = useState<number>();
+
+  const calculateTxCostInUsd = useCallback(() => {
```
Comment (Jabher) [inline hooks/useTxCostInUsd.ts:18]: it can be simply defined in useEffect, not as callback

File: next.config.mjs [hunk 14]
```diff
@@ -26,15 +44,22 @@ export default {
...
     styledComponents: true,
   },
+  experimental: {
+    // Fixes a build error with importing Pure ESM modules, e.g. reef-knot
+    // Some docs are here:
+    // <https://github.com/vercel/next.js/pull/27069>
+    // You can see how it is actually used in v12.3.4 here:
+    // <https://github.com/vercel/next.js/blob/v12.3.4/packages/next/build/webpack-config.ts#L417>
+    // Presumably, it is true by default in next v13 and won't be needed
+    esmExternals: true,
```
Comment (Jabher) [inline next.config.mjs:54]: as we're moving to next 13 it will be useless, I guess

File: scripts/build-dynamics.mjs [hunk 15]
```diff
@@ -19,5 +20,5 @@ export default () => {
   const path = resolve('./public/runtime/window-env.js');
   ensureDirSync(dirname(path));
   writeFileSync(path, `window.__env__=${JSON.stringify(dynamics)}`);
-  console.log('created runtime files');
+  console.log('Created runtime files!');
```
Comment (Jabher) [inline scripts/build-dynamics.mjs:23]: why?

File: utils/axios/instance.ts [hunk 16]
```diff
@@ -0,0 +1,16 @@
+// eslint-disable-next-line import/no-extraneous-dependencies
+import Axios, { CreateAxiosDefaults } from 'axios';
+import { REQUEST_TIMEOUT } from 'config';
+
+declare module 'axios' {
+  export interface InternalAxiosRequestConfig {
+    metadata: any;
```
Comment (Jabher) [inline utils/axios/instance.ts:7]:
```suggestion
    metadata?: Record<string, unknown>;
```

File: config/matomoClickEvents.ts [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+  faqLidoLandingLink = 'faqLidoLandingLink',
+  faqExamplePageLocalLink = 'faqExamplePageLocalLink',
+  // PAY ATTENTION: Extra matomo events can be added here
+}
+
+export const MATOMO_CLICK_EVENTS: Record<
+  MATOMO_CLICK_EVENTS_TYPES,
+  MatomoEventType
+> = {
+  [MATOMO_CLICK_EVENTS_TYPES.submitStake]: [
```
Comment (Jabher) [inline config/matomoClickEvents.ts:19]: faq will be changed

File: config/matomoFaq.ts [hunk 18]
```diff
@@ -0,0 +1,6 @@
+import { MATOMO_CLICK_EVENTS } from './matomoClickEvents';
+
+export const matomoEventMap = new Map();
+
+matomoEventMap.set('https://lido.fi/', MATOMO_CLICK_EVENTS.faqLidoLandingLink);
```
Comment (Jabher) [inline config/matomoFaq.ts:5]: far logic will be changed in this file

**Comment by hexnickk4997** [comment]
Probably there will be another iteration ;)

**Pull request closed without merging**